File: compiler/src/main/java/com/arkoisystems/arkoicompiler/stage/parser/ArkoiParser.java
/*
 * Copyright © 2019-2020 ArkoiSystems (https://www.arkoisystems.com/) All Rights Reserved.
 * Created ArkoiCompiler on May 12, 2020
 * Author єхcsє#5543 aka timo
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 *
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.arkoisystems.arkoicompiler.stage.parser;

import com.arkoisystems.alt.api.IError;
import com.arkoisystems.alt.api.INode;
import com.arkoisystems.alt.api.IParser;
import com.arkoisystems.alt.api.annotations.EntryPoint;
import com.arkoisystems.alt.api.annotations.Rule;
import com.arkoisystems.arkoicompiler.stage.lexer.ArkoiLexer;
import lombok.Getter;
import lombok.NonNull;
import lombok.SneakyThrows;
import org.jetbrains.annotations.NotNull;

@Getter
public class ArkoiParser implements IParser<ArkoiLexer>
{
    
    @NotNull
    private final String parserEBNF;
    
    @NotNull
    private final ArkoiLexer lexer;
    
    @EntryPoint(method = "onProgram")
    @SneakyThrows
    public ArkoiParser(@NotNull final ArkoiLexer lexer) {
        this.lexer = lexer;
        
        this.parserEBNF = new String(ArkoiLexer.class.getResourceAsStream(
                "/grammar/parser.ebnf"
        ).readAllBytes());
    }
    
    @Rule(name = "program")
    public void onProgram(@NotNull final INode node) { }
    
    @Rule(name = "global_statement")
    public void onGlobalStatement(@NotNull final INode node) { }
    
    @Rule(name = "function_declaration")
    public void onFunctionDeclaration(@NotNull final INode node) { }
    
    @Rule(name = "variable_declaration")
    public void onVariableDeclaration(@NotNull final INode node) { }
    
    @Rule(name = "import_declaration")
    public void onImportDeclaration(@NotNull final INode node) { }
    
    @Rule(name = "return_declaration")
    public void onReturnDeclaration(@NotNull final INode node) { }
    
    @Rule(name = "function_return")
    public void onFunctionReturn(@NotNull final INode node) { }
    
    @Rule(name = "function_arguments")
    public void onFunctionArguments(@NotNull final INode node) { }
    
    @Rule(name = "function_argument")
    public void onFunctionArgument(@NotNull final INode node) { }
    
    @Rule(name = "block")
    public void onBlock(@NotNull final INode node) { }
    
    @Rule(name = "inline_block")
    public void onInline_block(@NotNull final INode node) { }
    
    @Rule(name = "brace_block")
    public void onBraceBlock(@NotNull final INode node) { }
    
    @Rule(name = "block_statement")
    public void onBlockStatement(@NotNull final INode node) { }
    
    @Rule(name = "expression")
    public void onExpression(@NotNull final INode node) { }
    
    @Rule(name = "assignment_expression")
    public void onAssignmentExpression(@NotNull final INode node) { }
    
    @Rule(name = "additive_expression")
    public void onAdditiveExpression(@NotNull final INode node) { }
    
    @Rule(name = "multiplicative_expression")
    public void onMultiplicativeExpression(@NotNull final INode node) { }
    
    @Rule(name = "exponential_expression")
    public void onExponentialExpression(@NotNull final INode node) { }
    
    @Rule(name = "operable_expression")
    public void onOperableExpression(@NotNull final INode node) { }
    
    @Rule(name = "prefix_expression")
    public void onPrefixExpression(@NotNull final INode node) { }
    
    @Rule(name = "postfix_expression")
    public void onPostfixExpression(@NotNull final INode node) { }
    
    @Rule(name = "parenthesized_expression")
    public void onParenthesizedExpression(@NotNull final INode node) { }
    
    @Rule(name = "type")
    public void onType(@NotNull final INode node) { }
    
    @Override
    public void onError(final @NonNull IError error) {
        System.out.println("Arkoi Parser: " + error);
    }
    
}

File: ArkoiIntegration/src/main/kotlin/com/arkoisystems/lang/arkoi/parser/ArkoiParser.kt
/*
 * Copyright © 2019-2020 ArkoiSystems (https://www.arkoisystems.com/) All Rights Reserved.
 * Created ArkoiIntegration on March 27, 2020
 * Author єхcsє#5543 aka timo
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 *
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.arkoisystems.lang.arkoi.parser

import com.arkoisystems.arkoicompiler.ArkoiClass
import com.arkoisystems.arkoicompiler.api.ICompilerMarker
import com.arkoisystems.arkoicompiler.stage.syntaxAnalyzer.ast.utils.ASTType
import com.arkoisystems.arkoicompiler.stage.syntaxAnalyzer.marker.MarkerFactory
import com.arkoisystems.lang.arkoi.lexer.ArkoiLexer
import com.intellij.lang.ASTNode
import com.intellij.lang.LightPsiParser
import com.intellij.lang.PsiBuilder
import com.intellij.lang.PsiParser
import com.intellij.lang.impl.PsiBuilderImpl
import com.intellij.psi.tree.IElementType
import java.lang.reflect.Field

class ArkoiParser(private val arkoiParserDefinition: ArkoiParserDefinition) : PsiParser, LightPsiParser {

    private lateinit var myCurrentLexeme: Field

    private lateinit var arkoiClass: ArkoiClass

    override fun parseLight(root: IElementType, builder: PsiBuilder) {
        if (builder !is PsiBuilderImpl || builder.lexer !is ArkoiLexer)
            return

        this.myCurrentLexeme = builder.javaClass.getDeclaredField("myCurrentLexeme")
        this.myCurrentLexeme.isAccessible = true

        val arkoiLexer = builder.lexer as ArkoiLexer
        this.arkoiClass = arkoiLexer.arkoiClass

        this.arkoiClass.content = builder.originalText.toString().toCharArray()
        this.arkoiClass.lexicalAnalyzer.processStage()
        this.arkoiClass.syntaxAnalyzer.processStage()

        val mark = builder.mark()
        makeTree(arkoiLexer, root, builder, this.arkoiClass.syntaxAnalyzer.rootAST.markerFactory)
        mark.done(root)
    }

    override fun parse(root: IElementType, builder: PsiBuilder): ASTNode {
        parseLight(root, builder)
        val fileNode = builder.treeBuilt
        this.arkoiParserDefinition.arkoiClasses[fileNode.firstChildNode.psi.hashCode()] = this.arkoiClass
        return fileNode
    }

    private fun makeTree(
        arkoiLexer: ArkoiLexer,
        root: IElementType,
        builder: PsiBuilderImpl,
        markerFactory: MarkerFactory<*, *, *>
    ) {
        this.setLexemeIndex(builder, arkoiLexer.tokens!!.indexOf(markerFactory.currentMarker.start))

        val mark = builder.mark()

        for (nextFactory in markerFactory.nextMarkerFactories)
            this.makeTree(arkoiLexer, root, builder, nextFactory)

        this.setLexemeIndex(builder, arkoiLexer.tokens!!.indexOf(markerFactory.currentMarker.end) + 1)
        if (markerFactory.currentMarker.errorMessage != null) {
//            println(String.format(markerFactory.currentMarker.errorMessage!!, markerFactory.currentMarker.errorArguments))
//            println(markerFactory.currentMarker.errorMessage!! + ", " + markerFactory.currentMarker.errorArguments?.contentToString())
            mark.error(markerFactory.currentMarker.errorMessage!!)
        } else mark.done(getArkoiType(markerFactory.currentMarker))
    }

    private fun setLexemeIndex(builder: PsiBuilderImpl, index: Int) = this.myCurrentLexeme.set(builder, index)

    private fun getArkoiType(compilerMarker: ICompilerMarker<*, *>): IElementType {
        return when (compilerMarker.astType) {
            ASTType.IMPORT -> ArkoiElementTypes.import
            ASTType.NUMBER -> ArkoiElementTypes.number
            ASTType.CAST_EXPRESSION -> ArkoiElementTypes.castExpression
            ASTType.TYPE -> ArkoiElementTypes.type
            ASTType.ANNOTATION -> ArkoiElementTypes.annotation
            ASTType.ARGUMENT -> ArkoiElementTypes.argument
            ASTType.ASSIGNMENT_EXPRESSION -> ArkoiElementTypes.assignmentExpression
            ASTType.BINARY_EXPRESSION -> ArkoiElementTypes.binaryExpression
            ASTType.BLOCK -> ArkoiElementTypes.block
            ASTType.COLLECTION -> ArkoiElementTypes.collection
            ASTType.EQUALITY_EXPRESSION -> ArkoiElementTypes.equalityExpression
            ASTType.FUNCTION_CALL_PART -> ArkoiElementTypes.functionCallPart
            ASTType.FUNCTION -> ArkoiElementTypes.function
            ASTType.IDENTIFIER_CALL -> ArkoiElementTypes.identifierCall
            ASTType.LOGICAL_EXPRESSION -> ArkoiElementTypes.logicalExpression
            ASTType.OPERABLE -> ArkoiElementTypes.operable
            ASTType.PARAMETER -> ArkoiElementTypes.parameter
            ASTType.PARENTHESIZED_EXPRESSION -> ArkoiElementTypes.parenthesizedExpression
            ASTType.POSTFIX_EXPRESSION -> ArkoiElementTypes.postfixExpression
            ASTType.PREFIX_EXPRESSION -> ArkoiElementTypes.prefixExpression
            ASTType.RELATIONAL_EXPRESSION -> ArkoiElementTypes.relationalExpression
            ASTType.RETURN -> ArkoiElementTypes.`return`
            ASTType.STATEMENT -> ArkoiElementTypes.statement
            ASTType.STRING -> ArkoiElementTypes.string
            ASTType.VARIABLE -> ArkoiElementTypes.variable
            ASTType.EXPRESSION -> ArkoiElementTypes.expression
            ASTType.PARAMETER_LIST -> ArkoiElementTypes.parameterList
            ASTType.ARGUMENT_LIST -> ArkoiElementTypes.argumentList
            ASTType.ROOT -> ArkoiElementTypes.root

            else -> TODO(compilerMarker.astType.name)
        }
    }

}